components/script/dom/node.rs

```diff
@@ -1116,48 +1127,64 @@ impl LayoutNodeHelpers for LayoutJS<Node> {
 // Iteration and traversal
 //
 
-pub struct NodeSiblingIterator {
+pub struct SiblingIterator<T: DerivedFrom<Node>> {
     current: Option<Root<Node>>,
+    phantom: PhantomData<T>,
 }
 
-impl Iterator for NodeSiblingIterator {
-    type Item = Root<Node>;
+impl<T> Iterator for SiblingIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
+    fn next(&mut self) -> Option<Self::Item> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
-        self.current = current.GetNextSibling();
-        Some(current)
+        while let Some(next_sibling) = current.GetNextSibling() {
```

Comment:
That all seems to make sense to me.

I also like nox's idea below about a trait for upcast and downcast methods on iterators instead of adding a type parameter - I think I'm gonna stay hands-off on this until a bit more discussion happens.